sdk/storage/azure-storage-common/src/storage_retry_policy.cpp

```diff
@@ -0,0 +1,44 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include "azure/storage/common/internal/storage_retry_policy.hpp"
+
+namespace Azure { namespace Storage { namespace _internal {
+
+  bool StorageRetryPolicy::ShouldRetryOnResponse(
+      Core::Http::RawResponse const& response,
+      Core::Http::Policies::RetryOptions const& retryOptions,
+      int32_t attempt,
+      std::chrono::milliseconds& retryAfter,
+      double jitterFactor) const
+  {
+    bool ret = RetryPolicy::ShouldRetryOnResponse(
+        response, retryOptions, attempt, retryAfter, jitterFactor);
+    if (ret)
+    {
+      return true;
+    }
+
+    if (attempt > m_options.MaxRetries)
+    {
+      return false;
+    }
+
+    if (static_cast<std::underlying_type_t<Core::Http::HttpStatusCode>>(response.GetStatusCode())
+        >= 400)
+    {
+      const auto& headers = response.GetHeaders();
+      auto ite = headers.find("x-ms-copy-source-status-code");
+      if (ite != headers.end())
+      {
+        auto innerStatusCode = static_cast<Core::Http::HttpStatusCode>(std::stoi(ite->second));
+        if (m_options.StatusCodes.find(innerStatusCode) != m_options.StatusCodes.end())
+        {
+          return true;
+        }
```

Comment:
What's this inner status code check? It's not clear to me what we are trying to do here.

And do we retry for any status code?

Reply:
This is mainly for copy-operations.

For copy operations, client side sends an HTTP request to storage service to trigger this copy, and storage service will send one or more HTTP requests to source endpoint to do the actual work.
`innerStatusCode` here refers to the status code of HTTP requests between storage service and copy source endpoint.

Reply:
> And do we retry for any status code?

No, we only retry on the status codes specified by the customer. If not specified, these default values will be used.
https://github.com/Azure/azure-sdk-for-cpp/blob/main/sdk/core/azure-core/inc/azure/core/http/policies/policy.hpp#L106~L111